File: app/schemas/datasource.py
from typing import Optional, Dict, Any, List
from enum import Enum
from pydantic import BaseModel, Field, field_validator
from datetime import datetime


class SourceType(str, Enum):
    FILE = "file"
    API = "api"
    DATABASE = "database"
    URL = "url"


class ConnectionConfig(BaseModel):
    """Base model for connection details that will be extended by specific source types"""
    pass


class FileConnectionConfig(ConnectionConfig):
    """Configuration for file data sources"""
    file_path: Optional[str] = None
    file_format: str = Field(..., description="Format of the file (e.g., csv, json, txt, pdf)")
    encoding: str = "utf-8"
    delimiter: Optional[str] = None  # For CSV files


class ApiConnectionConfig(ConnectionConfig):
    """Configuration for API data sources"""
    url: str
    method: str = "GET"
    headers: Optional[Dict[str, str]] = None
    query_params: Optional[Dict[str, Any]] = None
    body: Optional[Dict[str, Any]] = None
    auth_type: Optional[str] = None
    auth_config: Optional[Dict[str, Any]] = None


class DatabaseConnectionConfig(ConnectionConfig):
    """Configuration for database data sources"""
    host: str
    port: int
    database: str
    username: str
    password: Optional[str] = None
    db_type: str = Field(..., description="Type of database (e.g., postgres, mysql, mssql)")
    ssl: bool = False
    query: Optional[str] = None
    table: Optional[str] = None


class UrlConnectionConfig(ConnectionConfig):
    """Configuration for URL data sources"""
    url: str
    scrape_config: Optional[Dict[str, Any]] = None
    headers: Optional[Dict[str, str]] = None


class DataSourceBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    source_type: SourceType
    connection_details: Dict[str, Any] = Field(..., description="Connection details specific to the source type")
    
    @field_validator('connection_details')
    def validate_connection_details(cls, v, values):
        source_type = values.get('source_type')
        if source_type == SourceType.FILE:
            FileConnectionConfig(**v)
        elif source_type == SourceType.API:
            ApiConnectionConfig(**v)
        elif source_type == SourceType.DATABASE:
            DatabaseConnectionConfig(**v)
        elif source_type == SourceType.URL:
            UrlConnectionConfig(**v)
        return v


class DataSourceCreate(DataSourceBase):
    credentials: Optional[Dict[str, Any]] = None


class DataSourceUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    connection_details: Optional[Dict[str, Any]] = None
    credentials: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class DataSource(DataSourceBase):
    id: str
    is_active: bool = True
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class FileUploadBase(BaseModel):
    data_source_id: str
    file_name: str
    file_type: str
    file_size: int
    storage_path: str


class FileUploadCreate(FileUploadBase):
    metadata: Optional[Dict[str, Any]] = None


class FileUploadStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class FileUpload(FileUploadBase):
    id: str
    upload_status: FileUploadStatus = FileUploadStatus.PENDING
    processed: bool = False
    metadata: Optional[Dict[str, Any]] = None
    uploaded_by: Optional[str] = None
    uploaded_at: datetime
    
    class Config:
        from_attributes = True

File: frontend/packages/schemas/src/datasource.ts
import { SourceType } from "@vbkg/types";
import { z } from "zod";
import { jsonSchema } from "./common";

const ConnectionDetailsSchema = z.object({
  name: z.string().min(1), // Name must be a non-empty string
  description: z.string().optional(), // Description is optional
  credentials: z
    .array(
      z.object({
        key: z.string().min(1), // Key must be a non-empty string
        value: z.string().min(1), // Value must be a non-empty string
      }),
    )
    .optional(), // Credentials are optional
});

const _FileConnectionDetailsSchema = z
  .object({
    source_type: z.literal(SourceType.FILE), // Type must be "file"
    connection_details: z.object({
      file_path: z.string(), // File path must be a string
      file_format: z
        .string()
        .describe("Format of the file (e.g., csv, json, txt, pdf)"),
      encoding: z.string().optional(), // Optional encoding for the file
      delimiter: z.string().optional(), // Optional delimiter for the file
    }),
  })
  .extend(ConnectionDetailsSchema.shape); // Extend the base connection details schema

const _DatabaseConnectionDetailsSchema = z
  .object({
    source_type: z.literal(SourceType.DATABASE), // Type must be "database"
    connection_details: z.object({
      host: z.string(), // Host must be a string
      port: z.number().int(), // Port must be an integer
      database: z.string(), // Database name must be a string
      username: z.string(), // Username must be a string
      password: z.string().optional(), // Optional password for the database
      db_type: z.enum(["postgres", "mysql", "mssql"]), // Database type must be one of the specified enums
      ssl: z.boolean().default(false), // SSL connection option, default is false
      query: z.string().optional(), // Optional SQL query to execute
      table: z.string().optional(), // Optional table name to connect to
    }),
  })
  .extend(ConnectionDetailsSchema.shape); // Extend the base connection details schema

const _ApiConnectionDetailsSchema = z
  .object({
    source_type: z.literal(SourceType.API), // Type must be "api"
    connection_details: z.object({
      url: z.string().url(), // URL must be a valid URL
      method: z.enum(["GET", "POST", "PUT", "DELETE"]), // HTTP method must be one of the specified
      headers: z.record(z.string()).optional(), // Optional headers for the API request
      query_params: z.record(z.string()).optional(), // Optional query parameters for the API request
      body: jsonSchema.optional(), // Optional body for the API request
      auth_type: z.enum(["BASIC", "BEARER"]).optional(), // Optional authentication type
      auth_config: z.record(z.any()).optional(), // Optional authentication configuration
    }),
  })
  .extend(ConnectionDetailsSchema.shape); // Extend the base connection details schema

const _URLConnectionDetailsSchema = z
  .object({
    source_type: z.literal(SourceType.URL), // Type must be "url"
    connection_details: z.object({
      url: z.string().url(), // URL must be a valid URL
      scrape_config: z.record(z.any()).optional(), // Optional scraping configuration
      headers: z.record(z.string()).optional(), // Optional headers for the URL request
    }),
  })
  .extend(ConnectionDetailsSchema.shape); // Extend the base connection details schema

// Schema for creating a data source
export const CreateDataSourceSchema = z.discriminatedUnion("source_type", [
  _FileConnectionDetailsSchema,
  _DatabaseConnectionDetailsSchema,
  _ApiConnectionDetailsSchema,
  _URLConnectionDetailsSchema,
]);

// Schema for updating a data source
export const UpdateDataSourceSchema = z.object({
  name: z.string().min(1).optional(), // Name can be updated or omitted
  config: z.record(z.string(), z.any()).optional(), // Config can be updated or omitted
});

// Schema for reading data sources with optional filters
export const ReadDataSourcesSchema = z.object({
  skip: z.number().min(0).default(0), // Pagination offset
  limit: z.number().min(1).default(100), // Pagination limit
  source_type: z.string().optional(), // Optional filter for source type
  is_active: z.boolean().optional(), // Optional filter for active status
});

// Schema for retrieving a specific data source by ID
export const ReadDataSourceByIdSchema = z.object({
  datasource_id: z.string().uuid(), // ID must be a valid UUID
});

// Schema for file upload metadata
export const FileUploadMetadataSchema = z.object({
  metadata: z.string().optional(), // Optional JSON metadata as a string
});

// Schema for updating file status
export const UpdateFileStatusSchema = z.object({
  status: z.enum(["PENDING", "PROCESSING", "COMPLETED", "FAILED"]), // Example enum for file status
  processed: z.boolean().optional(), // Indicates if the file is processed
  error_message: z.string().optional(), // Optional error message
});

// Schema for retrieving file uploads with optional filters
export const ReadFileUploadsSchema = z.object({
  datasource_id: z.string().uuid().optional(), // Optional filter by datasource ID
  status: z.enum(["PENDING", "PROCESSING", "COMPLETED", "FAILED"]).optional(), // Optional filter for file status
  processed: z.boolean().optional(), // Optional filter for processed files
  skip: z.number().min(0).default(0), // Pagination offset
  limit: z.number().min(1).default(100), // Pagination limit
});

// Schema for retrieving a specific file upload by ID
export const ReadFileUploadByIdSchema = z.object({
  file_id: z.string().uuid(), // ID must be a valid UUID
});

// Schema for deleting a file upload
export const DeleteFileUploadSchema = z.object({
  file_id: z.string().uuid(), // ID must be a valid UUID
});

// Schema for retrieving file content
export const GetFileContentSchema = z.object({
  file_id: z.string().uuid(), // ID must be a valid UUID
});
